include/pingcap/coprocessor/Client.h

```diff
@@ -52,57 +54,119 @@ struct copTask
+        : cop_req(req_),
+          tasks(std::move(tasks_)),
+          cluster(cluster_),
+          concurrency(concurrency_),
+          cancelled(false),
+          log(&Logger::get("pingcap/coprocessor"))
     {}
 
-    // fetch all data.
-    Exception prepare()
+    ~ResponseIter()
     {
-        std::vector<std::thread> worker_threads;
-        for (auto it = tasks.begin(); it != tasks.end(); it++)
+        cancelled = true;
+        for (auto it = worker_threads.begin(); it != worker_threads.end(); it++)
         {
-            std::thread worker(&ResponseIter::handle_task, this, *it);
+            it->join();
+        }
+    }
+
+    // send all tasks.
+    void open()
+    {
+        unfinished_thread = concurrency;
+        for (int i = 0; i < concurrency; i++)
+        {
+            std::thread worker(&ResponseIter::thread, this);
             worker_threads.push_back(std::move(worker));
         }
         log->debug("coprocessor has " + std::to_string(tasks.size()) + " tasks.");
-        for (auto it = worker_threads.begin(); it != worker_threads.end(); it++)
+    }
+
+    std::pair<Result, bool> next()
+    {
+        std::unique_lock<std::mutex> lk(results_mutex);
+        cond_var.wait(lk, [this] { return unfinished_thread == 0 || cancelled || results.size() > 0; });
+        if (cancelled)
         {
-            it->join();
+            return std::make_pair(Result(), false);
+        }
+        if (results.size() > 0)
+        {
+            auto ret = std::make_pair(results.front(), true);
+            results.pop();
+            return ret;
+        }
+        else
+        {
+            return std::make_pair(Result(), false);
         }
-        return cop_error;
     }
 
-    std::pair<std::string, bool> next()
+private:
+    void thread()
     {
-        if (idx < results.size())
+        while (true)
         {
-            return std::make_pair(results[idx++], true);
+            if (cancelled)
+            {
+                log->information("cop task has been cancelled");
+                unfinished_thread--;
+                return;
+            }
+            std::unique_lock<std::mutex> lk(fetch_mutex);
+            if (tasks.size() == task_index)
+            {
+                unfinished_thread--;
+                return;
+            }
+            const copTask & task = tasks[task_index];
+            task_index++;
+            lk.unlock();
+            handle_task(task);
         }
-        return std::make_pair(std::string(), false);
     }
 
-private:
     std::vector<copTask> handle_task_impl(kv::Backoffer & bo, const copTask & task);
     void handle_task(const copTask & task);
 
     Request * cop_req;
+    size_t task_index = 0;
     std::vector<copTask> tasks;
-    size_t idx = 0;
+    std::vector<std::thread> worker_threads;
+
     kv::Cluster * cluster;
+    int concurrency;
 
     std::mutex results_mutex;
-    std::mutex error_mutex;
+    std::mutex fetch_mutex;
```

Comment:
better to rename it as `task_mutex` or `fetch_task_mutext`

Reply:
fix, PTAL